File: solutions/python/simple-cipher/1/simple_cipher.py
import string
import re

class Cipher:
    def __init__(self, key=None):
        if key is None:
            import random
            key = ''.join(random.choice(string.ascii_lowercase) for _ in range(100))
        if not key or not re.match(r'^[a-z]+$', key):
            raise ValueError("Invalid key")
        self.key = key

    def encode(self, text):
        return ''.join(
            chr((ord(c) - 97 + ord(self.key[i % len(self.key)]) - 97) % 26 + 97)
            for i, c in enumerate(text)
        )

    def decode(self, text):
        return ''.join(
            chr((ord(c) - 97 - (ord(self.key[i % len(self.key)]) - 97)) % 26 + 97)
            for i, c in enumerate(text)
        )

File: solutions/ruby/simple-cipher/1/simple_cipher.rb
class Cipher
  attr_reader :key

  def initialize(key = nil)
    @key = key || (0...100).map { ('a'.ord + rand(26)).chr }.join
    raise ArgumentError unless @key.match?(/\A[a-z]+\z/)
  end

  def encode(text)
    text.chars.each_with_index.map do |c, i|
      shift = @key[i % @key.length].ord - 'a'.ord
      ((c.ord - 'a'.ord + shift) % 26 + 'a'.ord).chr
    end.join
  end

  def decode(text)
    text.chars.each_with_index.map do |c, i|
      shift = @key[i % @key.length].ord - 'a'.ord
      ((c.ord - 'a'.ord - shift) % 26 + 'a'.ord).chr
    end.join
  end
end
